[2.1] Enhance TextureButton and TextureFrame with resize

Geequlim opened the pull request:
* Add stretch mode `STRETCH_KEEP_ASPECT_COVERED` like CSS `background-size: cover`.
* Make same resize behavior for TextureButton with TextureFrame
* ~~~Remove property 'scale' of TextureButton which is not required any more.~~~
* Add option ResizeMode for backwards compatibility

Geequlim committed:
Make same resize behavior for TextureButton with TextureFrame.
Remove property 'scale' of TextureButton which is not required any more.

Geequlim committed:
Add Keep Covered texture resize mode

Geequlim commented:
@akien-mga Is there any way to make the code works with 3.0 without another PR?

RandomShaper commented:
He seems to be off-line lately. But I believe he would ask you to submit a separate PR for master.

akien-mga commented:
I prefer to avoid API removals in the stable branch, as we don't want released games to break when updating the engine to 2.1.3. The PR overall looks fine, but if you could keep the scale property and maybe just change how it works to use the new feature, that would be better.

Geequlim commented:
Well that's good idea.
Maybe I can drop the `scale` property on the 3.0 branch with anothor PR right?

Geequlim committed:
Add scale property back for backwards compatibility
now we have a choice
Fix textureframe modulate doesn't work with STRETCH_KEEP_ASPECT_CENTERED and STRETCH_KEEP_ASPECT

Geequlim commented:
Now we change select the resize mode :)
![](https://cloud.githubusercontent.com/assets/6964556/22695441/c57778da-ed85-11e6-94be-d0da12cc5975.png)

Geequlim commented on scene/gui/texture_button.cpp:
```diff
@@ -153,16 +205,22 @@ void TextureButton::_bind_methods() {
 	ADD_PROPERTYNZ(PropertyInfo(Variant::OBJECT,"textures/disabled",PROPERTY_HINT_RESOURCE_TYPE,"Texture"), _SCS("set_disabled_texture"), _SCS("get_disabled_texture"));
 	ADD_PROPERTYNZ(PropertyInfo(Variant::OBJECT,"textures/focused",PROPERTY_HINT_RESOURCE_TYPE,"Texture"), _SCS("set_focused_texture"), _SCS("get_focused_texture"));
 	ADD_PROPERTYNZ(PropertyInfo(Variant::OBJECT,"textures/click_mask",PROPERTY_HINT_RESOURCE_TYPE,"BitMap"), _SCS("set_click_mask"), _SCS("get_click_mask")) ;
+	ADD_PROPERTYNO(PropertyInfo(Variant::INT, "params/resize_mode",PROPERTY_HINT_ENUM,"Scale,Stretch"), _SCS("set_resize_mode"),_SCS("get_resize_mode"));
```
@akien-mga @RandomShaper 
That is stange the property`params/resize_mode` works with editor but can't be save.

Geequlim commented:
That is really strange the proptery `paprams/resize_mode` works fine with editor but it can't be saved into packed scene file.
@reduz  Do you have any idea with this?
https://github.com/GodotExplorer/godot/blob/pr-resizable-texturebutton/scene/gui/texture_button.cpp#L217-L220

bojidar-bg commented:
@Geequlim (blind guess) Can it be something tricky with the property type? Also, does it happen for "stretch_mode" too?

Geequlim commented:
@bojidar-bg The `stretch_mode` does not have this issue.
The type of the property `paprams/resize_mode` is a enumeration value of `TextureButton::ResizeMode`  which I added the following line to make it possible to covert to `int`
```c++
VARIANT_ENUM_CAST( TextureButton::ResizeMode );
```
Did I still missing something for enumeration exporting?

bojidar-bg commented:
You didn't miss anything I know of.... what happens if you rename the property to something like `params/resize_test` (in case it is some weird hashing failure)

Geequlim commented:
@bojidar-bg Tested with `params/resize_type`, `params/stype` and `resize_mode`.
None of them can be saved.

Geequlim commented:
@bojidar-bg I have found the problem in the code that I shouldn't export the property with `ADD_PROPERTYNO`
@akien-mga Now the pr is ready to check an merge :)
